flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/sink/filesystem/Bucket.java

```diff
@@ -53,175 +58,260 @@
+		long inProgressFileCreationTime = Long.MAX_VALUE;
+
+		if (inProgressPart != null) {
+			inProgressResumable = inProgressPart.persist();
+			inProgressFileCreationTime = inProgressPart.getCreationTime();
 		}
-		return commitable;
+
+		return new BucketState<>(bucketId, bucketPath, inProgressFileCreationTime, inProgressResumable, pendingPartsPerCheckpoint);
 	}
 
-	public void dispose() {
-		if (currentPart != null) {
-			currentPart.dispose();
+	private void prepareBucketForCheckpointing(long checkpointId) throws IOException {
+		if (inProgressPart != null && rollingPolicy.shouldRollOnCheckpoint(inProgressPart)) {
+			if (LOG.isDebugEnabled()) {
+				LOG.debug("Subtask {} closing in-progress part file for bucket id={} on checkpoint.", subtaskIndex, bucketId);
+			}
+			closePartFile();
+		}
+
+		if (!pendingPartsForCurrentCheckpoint.isEmpty()) {
+			pendingPartsPerCheckpoint.put(checkpointId, pendingPartsForCurrentCheckpoint);
+			pendingPartsForCurrentCheckpoint = new ArrayList<>();
 		}
 	}
 
-	public void onCheckpointAcknowledgment(long checkpointId) throws IOException {
+	void onSuccessfulCompletionOfCheckpoint(long checkpointId) throws IOException {
 		Preconditions.checkNotNull(fsWriter);
 
 		Iterator<Map.Entry<Long, List<RecoverableWriter.CommitRecoverable>>> it =
-				pendingPerCheckpoint.entrySet().iterator();
+				pendingPartsPerCheckpoint.entrySet().iterator();
 
 		while (it.hasNext()) {
 			Map.Entry<Long, List<RecoverableWriter.CommitRecoverable>> entry = it.next();
+
 			if (entry.getKey() <= checkpointId) {
-				for (RecoverableWriter.CommitRecoverable commitable : entry.getValue()) {
-					fsWriter.recoverForCommit(commitable).commit();
+				for (RecoverableWriter.CommitRecoverable committable : entry.getValue()) {
+					fsWriter.recoverForCommit(committable).commit();
 				}
 				it.remove();
 			}
 		}
 	}
 
-	public BucketState<BucketID> onCheckpoint(long checkpointId) throws IOException {
-		RecoverableWriter.ResumeRecoverable resumable = null;
-		long creationTime = Long.MAX_VALUE;
-
-		if (currentPart != null) {
-			resumable = currentPart.persist();
-			creationTime = currentPart.getCreationTime();
-		}
-
-		if (!pending.isEmpty()) {
-			pendingPerCheckpoint.put(checkpointId, pending);
-			pending = new ArrayList<>();
+	void rollOnProcessingTimeIfNeeded(long timestamp) throws IOException {
+		if (inProgressPart != null && rollingPolicy.shouldRollOnProcessingTime(inProgressPart, timestamp)) {
+			if (LOG.isDebugEnabled()) {
+				LOG.debug("Subtask {} closing in-progress part file for bucket id={} due to timeout " +
+						"(in-progress file created @ {}, last updated @ {} and now is {}).",
+						subtaskIndex, bucketId, inProgressPart.getCreationTime(), inProgressPart.getLastUpdateTime(), timestamp);
+			}
+			closePartFile();
 		}
-		return new BucketState<>(bucketId, bucketPath, creationTime, resumable, pendingPerCheckpoint);
 	}
 
-	private Path getNewPartPath() {
+	private Path assembleNewPartPath() {
```

Comment:
move below closePartFile so it's closer to where it is used. We could even in-line it since it's only used once.